Android/PushTutorial/app/src/main/java/io/ably/example/androidpushexample/MainActivity.java

```diff
@@ -1,61 +1,174 @@
+import android.content.IntentFilter;
 import android.os.Bundle;
+import android.os.Handler;
+import android.os.Message;
 import android.support.annotation.Nullable;
+import android.support.v4.content.LocalBroadcastManager;
 import android.support.v7.app.AppCompatActivity;
 import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.TextView;
 
 import io.ably.lib.realtime.AblyRealtime;
+import io.ably.lib.realtime.CompletionListener;
 import io.ably.lib.realtime.ConnectionStateListener;
 import io.ably.lib.types.AblyException;
 import io.ably.lib.types.ClientOptions;
+import io.ably.lib.types.ErrorInfo;
+import io.ably.lib.util.IntentUtils;
 
 /**
  * Created by Amit S.
  */
 public class MainActivity extends AppCompatActivity {
 
+    public static final int SUCCESS = 0;
+    public static final int FAILURE = 1;
+    public static final int UPDATE_LOGS = 2;
+
+    public static final String STEP_1 = "Initialize Ably";
+    public static final String STEP_2 = "Activate Push";
+    public static final String STEP_3 = "Subscribe Push Channel";
+    public static final String STEP_4 = "Send Push";
+
+    public static final String TEST_PUSH_CHANNEL = "test_push_channel";
+    private TextView rollingLogs;
+    private Button stepsButton;
+    private StringBuilder logs = new StringBuilder();
     private AblyRealtime ablyRealtime;
 
+
+    private Handler handler = new Handler(new Handler.Callback() {
+        @Override
+        public boolean handleMessage(Message msg) {
+            switch (msg.what) {
+                case SUCCESS:
+                    stepsButton.setText((String) msg.obj);
+                    stepsButton.setEnabled(true);
+                    break;
+                case FAILURE:
+                    stepsButton.setEnabled(false);
+                    break;
+                case UPDATE_LOGS:
+                    rollingLogs.setText(logs.toString());
+                    break;
+            }
+            return false;
+        }
+    });
+
+    private BroadcastReceiver pushReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            ErrorInfo error = IntentUtils.getErrorInfo(intent);
+            if (error != null) {
+                logMessage("Error activating push service: " + error);
+                handler.sendMessage(handler.obtainMessage(FAILURE));
+                return;
+            }
+
+            logMessage("Device is now registered for push");
+            handler.sendMessage(handler.obtainMessage(SUCCESS, STEP_3));
+        }
+    };
+
     @Override
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
-
-        initAblyRuntime();
+        setContentView(R.layout.activity_main);
+        rollingLogs = findViewById(R.id.rolling_logs);
+        stepsButton = findViewById(R.id.steps);
     }
 
-    private void initAblyRuntime() {
```

Comment:
Is there a way to avoid this change? You've removed the `try` and now have lots of updated lines in this commit unnecessarily.